pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java

```diff
@@ -279,4 +281,39 @@ protected byte[] getBytes(int dictId) {
   protected byte[] getBuffer() {
     return new byte[_numBytesPerValue];
   }
+
+  public void getDictIds(List<String> values, IntSet dictIds, int inPredicateSparseThreshold,
+      int inPredicateSortThreshold) {
+    if (length() / values.size() > inPredicateSparseThreshold || values.size() < inPredicateSortThreshold) {
+      for (String value : values) {
```

Comment:
Probably a micro-optimization, but here if `values.size() > inPredicateSparseThreshold `, we know values is sorted, so if we could expose the binary search function to accept a lower bound, we can replace this with something like

```
long lastDictId = 0;
for (String value : values) {
        int dictId = indexOf(llastDictId, value);
        if (dictId >= 0) {
          dictIds.add(dictId);
          lastDictId = dictId;
        }
}
```
i.e., since values is sorted, the search space for a value can be reduced further.

The binary search can accept a new param to specify start idx

```
  protected int binarySearch(int lowIdx, long value) {
    int low = lowIdx;
    int high = _length - 1;
    while (low <= high) {
      int mid = (low + high) >>> 1;
      long midValue = _valueReader.getLong(mid);
      if (midValue < value) {
        low = mid + 1;
      } else if (midValue > value) {
        high = mid - 1;
      } else {
        return mid;
      }
    }
    return -(low + 1);
  }
  ```

Reply:
Infact, if we write another binSearch implementation, which returns the index closest (but not greated) to the value being searched (and another indicator of whether the value actually matched), each of these searches will necessarily reduce the search space for the next value in the sorted list.

Reply:
Agreed. We should definitely benchmark this. One more thing to consider is about the random lookup vs sequential read for in-memory or on-disk data access.

Considering the Constant of each operation, it's hard to tell which one will be fast given our current numbers, e.g. N>1000, M > 100k.

Then we need to compare: O(N) * O(1 of sequential read) * lg(M) * O(1 of random read) vs (O(N) + O(M)) * O(1 of sequential read)

It's a bit hard to test until we run it on a large scale data set.